src/types/frozenset.rs

```diff
@@ -97,46 +97,143 @@ impl PyFrozenSet {
+/// `arbitrary_self_types`.
+#[doc(alias = "PyFrozenSet")]
+pub trait PyFrozenSetMethods<'py> {
+    /// Returns the number of items in the set.
+    ///
+    /// This is equivalent to the Python expression `len(self)`.
+    fn len(&self) -> usize;
+
+    /// Checks if set is empty.
+    fn is_empty(&self) -> bool;
+
+    /// Determines if the set contains the specified key.
+    ///
+    /// This is equivalent to the Python expression `key in self`.
+    fn contains<K>(&self, key: K) -> PyResult<bool>
+    where
+        K: ToPyObject;
+
+    /// Returns an iterator of values in this set.
+    fn iter(&self) -> BoundFrozenSetIterator<'py>;
+}
+
+impl<'py> PyFrozenSetMethods<'py> for Bound<'py, PyFrozenSet> {
+    #[inline]
+    fn len(&self) -> usize {
+        unsafe { ffi::PySet_Size(self.as_ptr()) as usize }
+    }
+
+    fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    fn contains<K>(&self, key: K) -> PyResult<bool>
+    where
+        K: ToPyObject,
+    {
+        fn inner(frozenset: &Bound<'_, PyFrozenSet>, key: Bound<'_, PyAny>) -> PyResult<bool> {
             match unsafe { ffi::PySet_Contains(frozenset.as_ptr(), key.as_ptr()) } {
                 1 => Ok(true),
                 0 => Ok(false),
                 _ => Err(PyErr::fetch(frozenset.py())),
             }
         }
 
-        inner(self, key.to_object(self.py()))
+        let py = self.py();
+        inner(self, key.to_object(py).into_bound(py))
     }
 
-    /// Returns an iterator of values in this frozen set.
-    pub fn iter(&self) -> PyFrozenSetIterator<'_> {
-        IntoIterator::into_iter(self)
+    fn iter(&self) -> BoundFrozenSetIterator<'py> {
+        BoundFrozenSetIterator::new(self.clone())
+    }
+}
+
+/// PyO3 implementation of an iterator for a Python `frozenset` object.
+pub struct PyFrozenSetIterator<'py>(BoundFrozenSetIterator<'py>);
+
+impl<'py> Iterator for PyFrozenSetIterator<'py> {
+    type Item = &'py super::PyAny;
+
+    /// Advances the iterator and returns the next value.
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        self.0.next().map(Bound::into_gil_ref)
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.0.size_hint()
+    }
+}
+
+impl<'py> IntoIterator for &'py PyFrozenSet {
+    type Item = &'py PyAny;
+    type IntoIter = PyFrozenSetIterator<'py>;
+    /// Returns an iterator of values in this set.
+    fn into_iter(self) -> Self::IntoIter {
+        PyFrozenSetIterator(BoundFrozenSetIterator::new(
+            Bound::borrowed_from_gil_ref(&self).clone(),
+        ))
+    }
+}
+
+impl<'py> IntoIterator for Bound<'py, PyFrozenSet> {
+    type Item = Bound<'py, PyAny>;
+    type IntoIter = BoundFrozenSetIterator<'py>;
+
+    /// Returns an iterator of values in this set.
+    fn into_iter(self) -> Self::IntoIter {
+        BoundFrozenSetIterator::new(self)
+    }
+}
+
+impl<'py> IntoIterator for &'_ Bound<'py, PyFrozenSet> {
```

Comment:
Didn't we remove this impl elsewhere as well? Since it just clones the `Bound`, i.e. increases the reference count, this should also be something that user code can just write directly, i.e. `val.clone().into_iter()`?

Reply:
I feel like it was useful at some point, but I cannot reproduce that now. I've pushed a commit to remove this implementation for now (for all `Bound` types).